pydantic/schema.py

```diff
@@ -405,16 +405,20 @@ def field_type_schema(
         definitions.update(f_definitions)
         return {'type': 'array', 'uniqueItems': True, 'items': f_schema}, definitions
     elif field.shape is Shape.MAPPING:
+        dict_schema = {'type': 'object'}
+        regex = getattr(field.key_field.type_, 'regex', None)
         f_schema, f_definitions = field_singleton_schema(
             field, by_alias=by_alias, model_name_map=model_name_map, ref_prefix=ref_prefix
         )
         definitions.update(f_definitions)
-        if f_schema:
-            # The dict values are not simply Any
-            return {'type': 'object', 'additionalProperties': f_schema}, definitions
-        else:
-            # The dict values are Any, no need to declare it
-            return {'type': 'object'}, definitions
+        if regex:
+            # Dict keys have a regex pattern
+            # f_schema might be a schema or empty dict, add it either way
+            dict_schema['patternProperties'] = {regex.pattern: f_schema}
+        elif f_schema:
```

Comment:
Sorry, one more question.

Are these two cases really mutually exclusive? Eg. requiring `elif`.

maybe you might want both `patternProperties` and `additionalProperties`?

Reply:
Yep. I thought about the same while implementing it.

The schema that would go in `additionalProperties` ends up as the value of the `dict` at `patternProperties`. That `dict` has as key the regex pattern, and as value the schema for the `dict` value.


---

Also [from the spec](http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.5.6):

> Validation with "additionalProperties" applies only to the child values of instance names that do not match any names in "properties", and do not match any regular expression in "patternProperties".

Reply:
I can think of a case that would allow for `patternProperties` and `additionalProperties`, but it is rather complex.

```python
>>> from typing import Union, Dict, Any
>>> from pydantic import BaseModel, constr
>>>
>>> Identifier = constr(regex=r'^([a-zA-Z_][a-zA-Z0-9_]*)$')
>>> Slug = constr(regex=r'^([a-zA-Z_-][a-zA-Z0-9_-]*)$')
>>> Number = constr(regex=r'^\d+$')
>>>
>>> class IdentifierModel(BaseModel):
...     alias: str = ...
...
>>> class SlugModel(BaseModel):
...     alias: str = ...
...     value: str = ...
...
>>> class NumberModel(BaseModel):
...     alias: str = ...
...     value: int = ...
...
>>> class Model(BaseModel):
...     items: Union[Dict[Identifier, IdentifierModel], Dict[Slug, SlugModel], Dict[Number, NumberModel], Dict[str, Any]] = {}
...
>>> print(Model.schema_json(indent=2))
{
  "title": "Model",
  "type": "object",
  "properties": {
    "items": {
      "title": "Items",
      "default": {},
      "anyOf": [
        {
          "type": "object",
          "patternProperties": {
            "^([a-zA-Z_][a-zA-Z0-9_]*)$": {
              "$ref": "#/definitions/IdentifierModel"
            }
          }
        },
        {
          "type": "object",
          "patternProperties": {
            "^([a-zA-Z_-][a-zA-Z0-9_-]*)$": {
              "$ref": "#/definitions/SlugModel"
            }
          }
        },
        {
          "type": "object",
          "patternProperties": {
            "^\\d+$": {
              "$ref": "#/definitions/NumberModel"
            }
          }
        },
        {
          "type": "object"
        }
      ]
    }
  },
  "definitions": {
    "IdentifierModel": {
      "title": "IdentifierModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        }
      },
      "required": [
        "alias"
      ]
    },
    "SlugModel": {
      "title": "SlugModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        },
        "value": {
          "title": "Value",
          "type": "string"
        }
      },
      "required": [
        "alias",
        "value"
      ]
    },
    "NumberModel": {
      "title": "NumberModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        },
        "value": {
          "title": "Value",
          "type": "integer"
        }
      },
      "required": [
        "alias",
        "value"
      ]
    }
  }
}
# expected
{
  "title": "Model",
  "type": "object",
  "properties": {
    "items": {
      "title": "Items",
      "default": {},
      "patternProperties": {
        "^([a-zA-Z_][a-zA-Z0-9_]*)$": {
          "$ref": "#/definitions/IdentifierModel"
        },
        "^([a-zA-Z_-][a-zA-Z0-9_-]*)$": {
          "$ref": "#/definitions/SlugModel"
        },
        "^\\d+$": {
          "$ref": "#/definitions/NumberModel"
        }
      },
      "additionalProperties": {},
    }
  },
  "definitions": {
    "IdentifierModel": {
      "title": "IdentifierModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        }
      },
      "required": [
        "alias"
      ]
    },
    "SlugModel": {
      "title": "SlugModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        },
        "value": {
          "title": "Value",
          "type": "string"
        }
      },
      "required": [
        "alias",
        "value"
      ]
    },
    "NumberModel": {
      "title": "NumberModel",
      "type": "object",
      "properties": {
        "alias": {
          "title": "Alias",
          "type": "string"
        },
        "value": {
          "title": "Value",
          "type": "integer"
        }
      },
      "required": [
        "alias",
        "value"
      ]
    }
  }
}
```

Reply:
Good find, however I think this is too niche for now, if you think it's a problem @demosdemon please create a new issue to discuss.